### DISCUSSION TITLE
How to Build an example manually

### DISCUSSION THREAD

**Discussion by nonunknown777** [ggml-org/llama.cpp#3830] [Q&A]
I would like to build examples/simple manually using gcc after I did make, I tried:

`gcc simple.cpp -I../../common/ -I../../spm-headers`

but still getting error:
```
simple.cpp:(.text._ZNSt10_Head_baseILm0ENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEELb0EED2Ev[_ZNSt10_Head_baseILm0ENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEELb0EED5Ev]+0x14): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
/usr/bin/ld: /tmp/ccwyglmE.o: in function `std::__new_allocator<int>::allocate(unsigned long, void const*)':
simple.cpp:(.text._ZNSt15__new_allocatorIiE8allocateEmPKv[_ZNSt15__new_allocatorIiE8allocateEmPKv]+0x4b): undefined reference to `std::__throw_bad_array_new_length()'
/usr/bin/ld: simple.cpp:(.text._ZNSt15__new_allocatorIiE8allocateEmPKv[_ZNSt15__new_allocatorIiE8allocateEmPKv]+0x50): undefined reference to `std::__throw_bad_alloc()'
/usr/bin/ld: simple.cpp:(.text._ZNSt15__new_allocatorIiE8allocateEmPKv[_ZNSt15__new_allocatorIiE8allocateEmPKv]+0x60): undefined reference to `operator new(unsigned long)'
/usr/bin/ld: /tmp/ccwyglmE.o: in function `std::__new_allocator<std::__detail::_Hash_node_base*>::deallocate(std::__detail::_Hash_node_base**, unsigned long)':
simple.cpp:(.text._ZNSt15__new_allocatorIPNSt8__detail15_Hash_node_baseEE10deallocateEPS2_m[_ZNSt15__new_allocatorIPNSt8__detail15_Hash_node_baseEE10deallocateEPS2_m]+0x2b): undefined reference to `operator delete(void*, unsigned long)'
/usr/bin/ld: /tmp/ccwyglmE.o: in function `std::__new_allocator<std::__detail::_Hash_node<std::pair<int const, float>, false> >::deallocate(std::__detail::_Hash_node<std::pair<int const, float>, false>*, unsigned long)':
simple.cpp:(.text._ZNSt15__new_allocatorINSt8__detail10_Hash_nodeISt4pairIKifELb0EEEE10deallocateEPS5_m[_ZNSt15__new_allocatorINSt8__detail10_Hash_nodeISt4pairIKifELb0EEEE10deallocateEPS5_m]+0x2a): undefined reference to `operator delete(void*, unsigned long)'
/usr/bin/ld: /tmp/ccwyglmE.o:(.data.rel.local.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'
```

**Comment by KerfuffleV2** [comment]
The simplest thing would be to `make clean` then `make simple`. It should print out the commands it uses to build.

Since this is C++, you can't use the plain C compiler. You also need to compile/link with the common stuff (in `common/`) that the examples depend on, the `llama.cpp` source, and GGML. If you want to enable GPU support it's going to be a lot more complicated.

You'll need to compile and then link with at these additional modules:

    ggml.o llama.o common.o sampling.o grammar-parser.o k_quants.o ggml-alloc.o ggml-backend.o

Basically, it's going to be a pain. So why not let `make` handle it for you?

**Reply by nonunknown777** [reply]
I want to compile simple example separately because I want to bind with godot GDExtension, so I would like to be able to compile each example, then I can link as standalone!

**Reply by KerfuffleV2** [reply]
I'm not sure if you mean you want to compile the `simple` example _as_ a Godot extension. If so, you'll have to make considerable changes since it's meant to run as a standalone executable.

If you just want to link with an external library though, the easiest way would be to edit the entry in the Makefile. Let's say for example you want to add `-I/blah/godot-includes` so that it can find the Godot include files, `-L/blah/godot-libs` so it can find Godot libraries and `-lfun-godot-stuff` to make it actually link with `libfun-godot-stuff.a` or whatever. You can find this in the `Makefile`:

```makefile
simple: examples/simple/simple.cpp                            build-info.h ggml.o llama.o $(COMMON_DEPS) $(OBJS)
	$(CXX) $(CXXFLAGS) $(filter-out %.h,$^) -o $@ $(LDFLAGS)
```

and change it to

```makefile
simple: examples/simple/simple.cpp                            build-info.h ggml.o llama.o $(COMMON_DEPS) $(OBJS)
	$(CXX) $(CXXFLAGS) -I/blah/godot-includes $(filter-out %.h,$^) -o $@ $(LDFLAGS) -L/blah/godot-libs -lfun-godot-stuff
```

**Reply by nonunknown777** [reply]
it wouldnt be simpler if I compile simple.cpp as a static library and link later in my godot project?

**Reply by KerfuffleV2** [reply]
That doesn't avoid linking with the stuff I mentioned. You'd just be linking it into the static library in that case instead of an executable.

No matter what, you're going to need to link with the GGML, `llama.cpp`, common stuff somehow.
